src/mining_stats.ts

```diff
@@ -0,0 +1,103 @@
+import {setupWeb3, setupContracts, chainUrl} from './utils/setup_utils';
+import {printInfo, setupBar, printHelp, parseArgs, printSuccess} from './utils/dialog_utils';
+import {saveData} from './utils/file_utils';
+import {sortChronologically, convertWeiToAmber} from './utils/event_utils';
+import {constants} from 'ambrosus-node-contracts';
+
+const additionalOptions = [{
+  name: 'roundcount',
+  alias: 'r',
+  type: Number,
+  description: 'How many recent consensus rounds we want to scan'
+}];
+
+const miningStats = async (): Promise<void> => {
+  const options = parseArgs(additionalOptions);
+  const web3 = await setupWeb3(options.rpc || chainUrl(options.env));
+  if (!options.roundcount || options.roundcount < 0) {
+    printHelp({
+      header: 'Mining stats',
+      content: `Gathers info about latest Apollos' activity.`
+    });
+    return;
+  }
+
+  const {validatorSetWrapper, blockchainStateWrapper, rolesEventEmitterWrapper} = await setupContracts(web3, options.headcontract);
+  const currentBlockNumber = await blockchainStateWrapper.getCurrentBlockNumber();
+
+  printInfo(`Fetching ${currentBlockNumber} blocks`);
+  const nodeOnboardingEvents = await rolesEventEmitterWrapper.nodeOnboardings(0, currentBlockNumber);
+  const nodeRetirementEvents = await rolesEventEmitterWrapper.nodeRetirements(0, currentBlockNumber);
+  const gatheredEvents = [...nodeOnboardingEvents, ...nodeRetirementEvents];
+
+  printInfo(`Sorting and filtering events...`);
+  const sortedAndFilteredEvents = sortChronologically(gatheredEvents.filter((event) => event.returnValues.role === constants.APOLLO));
+
+  const totalEventsNumber = sortedAndFilteredEvents.length;
+  printInfo(`${totalEventsNumber} events successfully extracted`);
+
+  printInfo(`Processing events...`);
+  const eventsProgressBar = await setupBar(totalEventsNumber);
+
+  const nodesState = {};
+
+  for (let index = 0; index < totalEventsNumber; index++) {
+    const nodeAddress = sortedAndFilteredEvents[index].returnValues.nodeAddress;
+    switch (sortedAndFilteredEvents[index].event) {
+      case 'NodeOnboarded':
+        nodesState[nodeAddress] = {
+          nodeAddress,
+          blocksMined: 0,
+          validator: false,
+          onboarded: true,
+          deposit: convertWeiToAmber(sortedAndFilteredEvents[index].returnValues.placedDeposit)
+        };
+        break;
+      case 'NodeRetired':
+        delete nodesState[nodeAddress];
+        break;
+    }
+    eventsProgressBar.increment(1);
+  }
+
+  printInfo(`Checking current validator set...`);
+  const validatorSet = await validatorSetWrapper.getValidators();
+
+  const nodeStateCheckedWithValidatorSet = validatorSet.reduce((nodes, current) => {
+    if (nodes[current] !== undefined) {
+      nodes[current].validator = true;
+    } else {
+    nodes[current] = {
+      nodeAddress: current,
+      blocksMined: 0,
+      onboarded: false,
+      validator: true};
+    }
+    return nodes;
+  }, nodesState);
```

Comment:
I don't really get why you need both parsing events and check validatorSet. We have following cases:
* Onboarded as validator - is in validatorSet
* Onboarded as not validator - we don't care
* Was onboarded as validator but is now retired - could be useful, but we delete all information 'bout 'em atm, also could be obtained as block miners not present in the VS
Only info we get from events is deposit but it'd simplify the code a lot if we got it by calling contract.
So maybe we don't need to parse events?